## Agenda

**Section 1 - A quick topic summary**

- Introduction to Numerical Computing
- Introduction to NumPy
- Creating and Manipulating NumPy Arrays
- NumPy Operations, Indexing
- Mathematical Functions and Matrix Operations in NumPy
- Introduction to Pandas and DataFrames
- Creating, Accessing, and Modifying DataFrames
- Data Filtering, Sorting, and Handling Missing Values
- Data Cleaning and Missing Value Treatment

**Section 2**  
- 5 objective questions

**Section 3**  

- Scenario-based practice questions

- An open Q&A block

**Closing Note**

## **1. Introduction to Numerical Computing**

Numerical computing means: using computers to solve mathematical and numerical problems.

It is widely used in:

- Artificial Intelligence
- Machine Learning
- Deep Learning
- Scientific Research
- Data Science
- Robotics
- Engineering
- Image Processing

**Why Numerical Computing?**

Computers can:

- perform calculations very fast
- process huge datasets
- solve complex mathematical problems
- automate computations

**Example:**

Instead of manually calculating:

1 + 2 + 3 + ... + 1000000

**Python can compute it instantly.**

## **2. Numerical Computing in Python**

Python provides powerful libraries:

**Library**   	    **Purpose**
-  **NumPy**	  ->      Numerical arrays and matrix operations
- **Pandas**	  ->      Data analysis and DataFrames
- **Matplotlib** -> 	  Data visualization
- **SciPy**	       ->   Scientific computing
- **PyTorch/TensorFlow**  ->	Deep learning

## **3. Why NumPy and Pandas?**

**NumPy**

Used for:

- fast numerical computation
- arrays
- matrix operations


**Pandas**

Used for:

- tabular data
- CSV files
- data cleaning
- data analysis

**Simple Example Without NumPy:**



In [ ]:
numbers = [1,2,3,4]

result = []

for i in numbers:
    result.append(i*2)

print(result)

[2, 4, 6, 8]


**With NumPy:**

In [ ]:
import numpy as np

arr = np.array([1,2,3,4])

print(arr*2)

[2 4 6 8]


**NumPy is:** shorter, faster, easier

## **4. What is NumPy?**

NumPy stands for: **Num**erical **Py**thon (Numpy)

It is a **Python library** **used for:**
- numerical computations
- matrix operations
- array processing
- scientific computing
- machine learning

**Why NumPy is Important?**

**NumPy is:**

- faster than Python lists
- memory efficient
- easy for mathematical operations
- widely used in AI and Data Science

## **Problem with Python Lists**

Suppose we want to multiply all numbers by 2.

In [ ]:
# Using python list

numbers = [1,2,3,4]

result = []

for i in numbers:
    result.append(i*2)

print(result)


[2, 4, 6, 8]


**This works, but:**

- requires loops
- slower for large data

In [ ]:
# Using numpy library

import numpy as np

arr = np.array([1,2,3,4])

print(arr * 2)

[2 4 6 8]


## **5. Installing NumPy**

In [ ]:
!pip install numpy

**Importing NumPy**

In [ ]:
import numpy as np

- **numpy** -> 	library name
- **np**	- > shortcut alias

## **6. NumPy Array**

A NumPy array is: a **collection of elements** stored efficiently in memory.

- **Arrays can be:** 1D, 2D, 3D, multi-dimensional

## **Creating 1D Array**

In [ ]:
import numpy as np

arr = np.array([1,2,3,4,5])

print(arr)

[1 2 3 4 5]


## **Creating 2D Array**

In [ ]:
arr2 = np.array([
    [1,2,3],
    [4,5,6]
])

print(arr2)

[[1 2 3]
 [4 5 6]]


## **Difference Between List and Array**

**Python List**
- slower
- more memory
- limited math operations
- not optimized

**NumPy Array**
- faster
- less memory
- powerful operations
- optimized

**Array Properties**

- shape - rows and columns
- ndim - 	number of dimensions
- size - 	total elements
- dtype - 	data type

In [ ]:
arr = np.array([
    [1,2,3],
    [4,5,6]
])

print(arr.shape,"Shape") #
print(arr.ndim, "Dim")
print(arr.size,"Size")
print(arr.dtype,"Data Type")

(2, 3) Shape
2 Dim
6 Size
int64 Data Type


**Creating Array of Zeros**

In [ ]:
zeros_array = np.zeros((3,3))

print(zeros_array)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


- creates matrix filled with zeros
- (3,3) means 3 rows and 3 columns

**Creating Array of Ones**

In [ ]:
ones_array = np.ones((2,4))

print(ones_array)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]]


**Identity Matrix**

- diagonal values = 1
- remaining values = 0

In [ ]:
identity = np.eye(4)

print(identity)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


**Creates array with range of values.**

In [ ]:
arr = np.arange(0,10)

print(arr)

[0 1 2 3 4 5 6 7 8 9]


**Using Step Size**

In [ ]:
arr = np.arange(0,20,2)

print(arr)

[ 0  2  4  6  8 10 12 14 16 18]


## **Random Arrays**

In [ ]:
# Random Decimal Values
arr = np.random.rand(3,3)

print(arr)

[[0.20803828 0.00609155 0.92217623]
 [0.91904574 0.67470411 0.70175029]
 [0.83199145 0.19827321 0.21828067]]


In [ ]:
# Random Integer Values
arr = np.random.randint(0,10,(2,3))

print(arr)

[[0 7 4]
 [5 0 3]]


In [ ]:
# Array Shape
arr = np.array([
    [1,2,3],
    [4,5,6]
])

print(arr.shape)

(2, 3)


In [ ]:
#Reshaping Arrays -> Reshape changes dimensions.
arr = np.arange(12)

print(arr)

[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [ ]:
# Reshape into Matrix
arr = arr.reshape(3,4)

print(arr)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [ ]:
# Flatten Array -> Convert multi-dimensional array to 1D.
arr = np.array([
    [1,2],
    [3,4]
])

flat = arr.flatten()

print(flat)

[1 2 3 4]


In [ ]:
# Copying Arrays
arr1 = np.array([1,2,3])

arr2 = arr1.copy()

print(arr2)

[1 2 3]


In [ ]:
# Changing Data Type
arr = np.array([1,2,3])

arr = arr.astype(float)

print(arr)

[1. 2. 3.]


In [ ]:
# Checking Array Size
arr = np.array([
    [1,2,3],
    [4,5,6]
])

print(arr.size)

6


In [ ]:
# Checking Number of Dimensions
print(arr.ndim)

2


## **7. NumPy Operations, Indexing**

**NumPy Arithmetic Operations** - NumPy allows mathematical operations directly on arrays.

In [ ]:
# Addition
import numpy as np

arr1 = np.array([1,2,3])

arr2 = np.array([4,5,6])

print(arr1 + arr2)

[5 7 9]


In [ ]:
# Subtraction
print(arr1 - arr2)


[-3 -3 -3]


In [ ]:
# Multiplication - Element-wise multiplication.

print(arr1 * arr2)

[ 4 10 18]


In [ ]:
# Division

print(arr1 / arr2)

[0.25 0.4  0.5 ]


In [ ]:
# Scalar Operations - Operations with single values.
arr = np.array([1,2,3])

print(arr + 10)

[11 12 13]


**NumPy Vectorization**

**Vectorization means:** performing operations without loops.

In [ ]:
# Traditional Python:
numbers = [1,2,3]

result = []

for i in numbers:
    result.append(i*2)

print(result)

[2, 4, 6]


In [ ]:
# Using Numpy : Advantages: faster, shorter code, optimized
arr = np.array([1,2,3])

print(arr * 2)

[2 4 6]


## **Indexing Arrays**

**Indexing means**: accessing specific elements.

In [ ]:
# 1D Indexing
arr = np.array([10,20,30,40,50])

print(arr[0])

10


In [ ]:
# Negative Indexing
print(arr[-1])  # -1 means last element

50


## **Mathematical Functions in NumPy**
- NumPy provides many built-in mathematical functions.

In [ ]:
#Creating Sample Array
import numpy as np

arr = np.array([1,2,3,4,5])

print(arr)

[1 2 3 4 5]


In [ ]:
# sum of elements
print(np.sum(arr))

15


In [ ]:
# mean (average)
print(np.mean(arr))

3.0


In [ ]:
# maximum value
print(np.max(arr))

5


In [ ]:
# minimum value
print(np.min(arr))

1


In [ ]:
# standard deviation
print(np.std(arr))

1.4142135623730951


In [ ]:
# variance
print(np.var(arr))

2.0


In [ ]:
# square root
arr = np.array([1,4,9,16])

print(np.sqrt(arr))

[1. 2. 3. 4.]


## **7. Introduction to Pandas and DataFrames**

**What is Pandas?**

Pandas is a Python library used for:

- data analysis
- data manipulation
- handling tabular data
- CSV file processing
- data cleaning

It is **one of the most important libraries** in:

- Data Science
- Machine Learning
- Artificial Intelligence

**Why Pandas?**

**Pandas helps us** to:

- organize data
- analyze data
- clean missing values
- filter data
- sort data
- perform statistical analysis

## **Installing Pandas**

In [ ]:
!pip install pandas

**Importing Pandas**

In [ ]:
import pandas as pd
# pandas is library name
# pd is shortcut alias

**What is a DataFrame?**

A **DataFrame** is: a **table of rows and columns**.

| Name    | Age | Marks |
| ------- | --- | ----- |
| Alice   | 20  | 85    |
| Bob     | 21  | 90    |
| Charlie | 22  | 95    |


**Real-Life Examples of DataFrames**
- Student records
- Excel sheets
- CSV files
- Employee databases
- Sales reports

**What is a Series?**

A **Series** is: **a single column** of data.

In [ ]:
#Example
import pandas as pd

s = pd.Series([10,20,30,40])

print(s)

0    10
1    20
2    30
3    40
dtype: int64


## **Creating a DataFrame**
Method 1 — Using Dictionary

In [ ]:
data = {

    'Name':['Alice','Bob','Charlie'],

    'Age':[20,21,22],

    'Marks':[85,90,95]
}

df = pd.DataFrame(data)

print(df)

      Name  Age  Marks
0    Alice   20     85
1      Bob   21     90
2  Charlie   22     95


**Dataframe Cmponents**

| Component | Meaning         |
| --------- | --------------- |
| rows      | horizontal data |
| columns   | vertical data   |
| index     | row numbers     |


**Checking DataFrame Information**

In [ ]:
print(df.shape) # shape 3 rows, 3 columns

(3, 3)


In [ ]:
# column name
print(df.columns)

Index(['Name', 'Age', 'Marks'], dtype='object')


In [ ]:
# Data type
print(df.dtypes)

Name     object
Age       int64
Marks     int64
dtype: object


In [ ]:
# summary information
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    3 non-null      object
 1   Age     3 non-null      int64 
 2   Marks   3 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 204.0+ bytes
None


In [ ]:
# Statistical Summary
print(df.describe())

        Age  Marks
count   3.0    3.0
mean   21.0   90.0
std     1.0    5.0
min    20.0   85.0
25%    20.5   87.5
50%    21.0   90.0
75%    21.5   92.5
max    22.0   95.0


## **Reading CSV(Comma Separated Values) File**

**CSV example**

Name,Age,Marks

Alice,20,85

Bob,21,90

**Loading CSV File**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

KeyboardInterrupt: 

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students_dataset.csv')

**Viewing Top Rows**

In [ ]:
print(df.head())

   Student_ID     Name   Age Department  Marks  Attendance     City Passed
0         101    Alice  20.0        CSE   85.0        90.0    Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0   Mumbai    Yes
2         103  Charlie  22.0        CSE    NaN        88.0    Delhi    Yes
3         104    David   NaN         ME   78.0         NaN  Chennai     No
4         105     Emma  20.0        ECE   92.0        95.0     Pune    Yes


**Viewing Specific Number of Rows**

In [ ]:
print(df.head(2))

   Student_ID   Name   Age Department  Marks  Attendance    City Passed
0         101  Alice  20.0        CSE   85.0        90.0   Delhi    Yes
1         102    Bob  21.0        ECE   90.0        85.0  Mumbai    Yes


**Viewing Last Rows**

In [ ]:
print(df.tail())

   Student_ID   Name   Age Department  Marks  Attendance       City Passed
5         106  Frank  23.0        CSE   88.0        80.0      Delhi    Yes
6         107  Grace  21.0         ME   67.0        70.0  Bangalore     No
7         108  Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
8         109    Ian   NaN        CSE   73.0        75.0        NaN     No
9         110   Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


## Accessing Columns

**Single column**

In [ ]:
print(df['Name'])

0      Alice
1        Bob
2    Charlie
3      David
4       Emma
5      Frank
6      Grace
7      Helen
8        Ian
9       Jack
Name: Name, dtype: object


**Multiple Columns**

In [ ]:
print(df[['Name','Marks']])

      Name  Marks
0    Alice   85.0
1      Bob   90.0
2  Charlie    0.0
3    David   78.0
4     Emma   92.0
5    Frank   88.0
6    Grace   67.0
7    Helen   95.0
8      Ian   73.0
9     Jack   81.0


## **Accessing Rows Using iloc**

In [ ]:
print(df.iloc[0])

Student_ID      101
Name          Alice
Age            20.0
Department      CSE
Marks          85.0
Attendance     90.0
City          Delhi
Passed          Yes
Name: 0, dtype: object


**Multiple Rows**

In [ ]:
print(df.iloc[0:3])

   Student_ID     Name   Age Department  Marks  Attendance    City Passed
0         101    Alice  20.0        CSE   85.0        90.0   Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0  Mumbai    Yes
2         103  Charlie  22.0        CSE    0.0        88.0   Delhi    Yes


**Access Specific Value**

In [ ]:
print(df.iloc[0,2])

# Using loc
print(df.loc[0]) # [row_index, column_index]

20.0
Student_ID      101
Name          Alice
Age            20.0
Department      CSE
Marks          85.0
Attendance     90.0
City          Delhi
Passed          Yes
Name: 0, dtype: object


**Access Specific Column with loc**

In [ ]:
print(df.loc[:, 'Marks']) # all rows, Marks column

0    85.0
1    90.0
2     0.0
3    78.0
4    92.0
5    88.0
6    67.0
7    95.0
8    73.0
9    81.0
Name: Marks, dtype: float64


**Checking Shape**

In [ ]:
print(df.shape) # (rows, columns)

## **Access Marks Column**

In [ ]:
print(df['Marks'])

0    85.0
1    90.0
2     NaN
3    78.0
4    92.0
5    88.0
6    67.0
7    95.0
8    73.0
9    81.0
Name: Marks, dtype: float64


## **Update Missing Marks**

In [ ]:
df['Marks'] = df['Marks'].fillna(0)

## **Delete City Column**

In [ ]:
df.drop('City', axis=1)

,Student_ID,Name,Age,Department,Marks,Attendance,Passed
0,101,Alice,20.0,CSE,85.0,90.0,Yes
1,102,Bob,21.0,ECE,90.0,85.0,Yes
2,103,Charlie,22.0,CSE,0.0,88.0,Yes
3,104,David,NaN,ME,78.0,NaN,No
4,105,Emma,20.0,ECE,92.0,95.0,Yes
5,106,Frank,23.0,CSE,88.0,80.0,Yes
6,107,Grace,21.0,ME,67.0,70.0,No
7,108,Helen,22.0,ECE,95.0,92.0,Yes
8,109,Ian,NaN,CSE,73.0,75.0,No
9,110,Jack,24.0,NaN,81.0,89.0,Yes


## **Modifying DataFrames**

In [ ]:
# Updating Single Value
df.loc[0,'Marks'] = 100

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
0         101    Alice  20.0        CSE  100.0        90.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
2         103  Charlie  22.0        CSE    0.0        88.0      Delhi    Yes
3         104    David   NaN         ME   78.0         NaN    Chennai     No
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


In [ ]:
# Updating Entire Colum
df['Marks'] = df['Marks'] + 5   # Adds 5 marks to every student.

In [ ]:
# Adding New Column
df['Result'] = 'Pass'

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed  \
0         101    Alice  20.0        CSE  110.0        90.0      Delhi    Yes   
1         102      Bob  21.0        ECE  100.0        85.0     Mumbai    Yes   
2         103  Charlie  22.0        CSE   10.0        88.0      Delhi    Yes   
3         104    David   NaN         ME   88.0         NaN    Chennai     No   
4         105     Emma  20.0        ECE  102.0        95.0       Pune    Yes   
5         106    Frank  23.0        CSE   98.0        80.0      Delhi    Yes   
6         107    Grace  21.0         ME   77.0        70.0  Bangalore     No   
7         108    Helen  22.0        ECE  105.0        92.0     Mumbai    Yes   
8         109      Ian   NaN        CSE   83.0        75.0        NaN     No   
9         110     Jack  24.0        NaN   91.0        89.0  Hyderabad    Yes   

  Result  
0   Pass  
1   Pass  
2   Pass  
3   Pass  
4   Pass  
5   Pass  
6   Pass  
7   Pass  
8   Pass  
9   Pass 

In [ ]:
# Conditional Column
df['Grade'] = [

    'A',
    'B',
    'A',
    'C',
    'A',
    'B',
    'C',
    'A',
    'B',
    'B'
]

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed  \
0         101    Alice  20.0        CSE  110.0        90.0      Delhi    Yes   
1         102      Bob  21.0        ECE  100.0        85.0     Mumbai    Yes   
2         103  Charlie  22.0        CSE   10.0        88.0      Delhi    Yes   
3         104    David   NaN         ME   88.0         NaN    Chennai     No   
4         105     Emma  20.0        ECE  102.0        95.0       Pune    Yes   
5         106    Frank  23.0        CSE   98.0        80.0      Delhi    Yes   
6         107    Grace  21.0         ME   77.0        70.0  Bangalore     No   
7         108    Helen  22.0        ECE  105.0        92.0     Mumbai    Yes   
8         109      Ian   NaN        CSE   83.0        75.0        NaN     No   
9         110     Jack  24.0        NaN   91.0        89.0  Hyderabad    Yes   

  Result Grade  
0   Pass     A  
1   Pass     B  
2   Pass     A  
3   Pass     C  
4   Pass     A  
5   Pass     B  


In [ ]:
# Creating Calculated Column
df['Bonus_Marks'] = df['Marks'] + 10

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed  \
0         101    Alice  20.0        CSE  110.0        90.0      Delhi    Yes   
1         102      Bob  21.0        ECE  100.0        85.0     Mumbai    Yes   
2         103  Charlie  22.0        CSE   10.0        88.0      Delhi    Yes   
3         104    David   NaN         ME   88.0         NaN    Chennai     No   
4         105     Emma  20.0        ECE  102.0        95.0       Pune    Yes   
5         106    Frank  23.0        CSE   98.0        80.0      Delhi    Yes   
6         107    Grace  21.0         ME   77.0        70.0  Bangalore     No   
7         108    Helen  22.0        ECE  105.0        92.0     Mumbai    Yes   
8         109      Ian   NaN        CSE   83.0        75.0        NaN     No   
9         110     Jack  24.0        NaN   91.0        89.0  Hyderabad    Yes   

  Result Grade  Bonus_Marks  
0   Pass     A        120.0  
1   Pass     B        110.0  
2   Pass     A         20.0  

In [ ]:
# Renaming Columns
df.rename(columns={'Marks':'Score'})

,Student_ID,Name,Age,Department,Score,Attendance,City,Passed,Result,Grade,Bonus_Marks
0,101,Alice,20.0,CSE,110.0,90.0,Delhi,Yes,Pass,A,120.0
1,102,Bob,21.0,ECE,100.0,85.0,Mumbai,Yes,Pass,B,110.0
2,103,Charlie,22.0,CSE,10.0,88.0,Delhi,Yes,Pass,A,20.0
3,104,David,NaN,ME,88.0,NaN,Chennai,No,Pass,C,98.0
4,105,Emma,20.0,ECE,102.0,95.0,Pune,Yes,Pass,A,112.0
5,106,Frank,23.0,CSE,98.0,80.0,Delhi,Yes,Pass,B,108.0
6,107,Grace,21.0,ME,77.0,70.0,Bangalore,No,Pass,C,87.0
7,108,Helen,22.0,ECE,105.0,92.0,Mumbai,Yes,Pass,A,115.0
8,109,Ian,NaN,CSE,83.0,75.0,NaN,No,Pass,B,93.0
9,110,Jack,24.0,NaN,91.0,89.0,Hyderabad,Yes,Pass,B,101.0


In [ ]:
# Deleting Column
df.drop('Grade', axis=1)

,Student_ID,Name,Age,Department,Marks,Attendance,City,Passed,Result,Bonus_Marks
0,101,Alice,20.0,CSE,110.0,90.0,Delhi,Yes,Pass,120.0
1,102,Bob,21.0,ECE,100.0,85.0,Mumbai,Yes,Pass,110.0
2,103,Charlie,22.0,CSE,10.0,88.0,Delhi,Yes,Pass,20.0
3,104,David,NaN,ME,88.0,NaN,Chennai,No,Pass,98.0
4,105,Emma,20.0,ECE,102.0,95.0,Pune,Yes,Pass,112.0
5,106,Frank,23.0,CSE,98.0,80.0,Delhi,Yes,Pass,108.0
6,107,Grace,21.0,ME,77.0,70.0,Bangalore,No,Pass,87.0
7,108,Helen,22.0,ECE,105.0,92.0,Mumbai,Yes,Pass,115.0
8,109,Ian,NaN,CSE,83.0,75.0,NaN,No,Pass,93.0
9,110,Jack,24.0,NaN,91.0,89.0,Hyderabad,Yes,Pass,101.0


In [ ]:
# Deleting Multiple Columns
df.drop(['Age','Result'], axis=1)

,Student_ID,Name,Department,Marks,Attendance,City,Passed,Grade,Bonus_Marks
0,101,Alice,CSE,110.0,90.0,Delhi,Yes,A,120.0
1,102,Bob,ECE,100.0,85.0,Mumbai,Yes,B,110.0
2,103,Charlie,CSE,10.0,88.0,Delhi,Yes,A,20.0
3,104,David,ME,88.0,NaN,Chennai,No,C,98.0
4,105,Emma,ECE,102.0,95.0,Pune,Yes,A,112.0
5,106,Frank,CSE,98.0,80.0,Delhi,Yes,B,108.0
6,107,Grace,ME,77.0,70.0,Bangalore,No,C,87.0
7,108,Helen,ECE,105.0,92.0,Mumbai,Yes,A,115.0
8,109,Ian,CSE,83.0,75.0,NaN,No,B,93.0
9,110,Jack,NaN,91.0,89.0,Hyderabad,Yes,B,101.0


In [ ]:
# Deleting Rows
df.drop(0) # delete first row

,Student_ID,Name,Age,Department,Marks,Attendance,City,Passed,Result,Grade,Bonus_Marks
1,102,Bob,21.0,ECE,100.0,85.0,Mumbai,Yes,Pass,B,110.0
2,103,Charlie,22.0,CSE,10.0,88.0,Delhi,Yes,Pass,A,20.0
3,104,David,NaN,ME,88.0,NaN,Chennai,No,Pass,C,98.0
4,105,Emma,20.0,ECE,102.0,95.0,Pune,Yes,Pass,A,112.0
5,106,Frank,23.0,CSE,98.0,80.0,Delhi,Yes,Pass,B,108.0
6,107,Grace,21.0,ME,77.0,70.0,Bangalore,No,Pass,C,87.0
7,108,Helen,22.0,ECE,105.0,92.0,Mumbai,Yes,Pass,A,115.0
8,109,Ian,NaN,CSE,83.0,75.0,NaN,No,Pass,B,93.0
9,110,Jack,24.0,NaN,91.0,89.0,Hyderabad,Yes,Pass,B,101.0


In [ ]:
# Setting Custom Index
df.set_index('Name')

,Student_ID,Age,Department,Marks,Attendance,City,Passed,Result,Grade,Bonus_Marks
Name,,,,,,,,,,
Alice,101,20.0,CSE,110.0,90.0,Delhi,Yes,Pass,A,120.0
Bob,102,21.0,ECE,100.0,85.0,Mumbai,Yes,Pass,B,110.0
Charlie,103,22.0,CSE,10.0,88.0,Delhi,Yes,Pass,A,20.0
David,104,NaN,ME,88.0,NaN,Chennai,No,Pass,C,98.0
Emma,105,20.0,ECE,102.0,95.0,Pune,Yes,Pass,A,112.0
Frank,106,23.0,CSE,98.0,80.0,Delhi,Yes,Pass,B,108.0
Grace,107,21.0,ME,77.0,70.0,Bangalore,No,Pass,C,87.0
Helen,108,22.0,ECE,105.0,92.0,Mumbai,Yes,Pass,A,115.0
Ian,109,NaN,CSE,83.0,75.0,NaN,No,Pass,B,93.0


In [ ]:
# Resetting Index
df.reset_index()

,index,Student_ID,Name,Age,Department,Marks,Attendance,City,Passed,Result,Grade,Bonus_Marks
0,0,101,Alice,20.0,CSE,110.0,90.0,Delhi,Yes,Pass,A,120.0
1,1,102,Bob,21.0,ECE,100.0,85.0,Mumbai,Yes,Pass,B,110.0
2,2,103,Charlie,22.0,CSE,10.0,88.0,Delhi,Yes,Pass,A,20.0
3,3,104,David,NaN,ME,88.0,NaN,Chennai,No,Pass,C,98.0
4,4,105,Emma,20.0,ECE,102.0,95.0,Pune,Yes,Pass,A,112.0
5,5,106,Frank,23.0,CSE,98.0,80.0,Delhi,Yes,Pass,B,108.0
6,6,107,Grace,21.0,ME,77.0,70.0,Bangalore,No,Pass,C,87.0
7,7,108,Helen,22.0,ECE,105.0,92.0,Mumbai,Yes,Pass,A,115.0
8,8,109,Ian,NaN,CSE,83.0,75.0,NaN,No,Pass,B,93.0
9,9,110,Jack,24.0,NaN,91.0,89.0,Hyderabad,Yes,Pass,B,101.0


In [ ]:
# Saving Modified DataFrame
df.to_csv('updated_students.csv', index=False)

# index=False, prevents saving row numbers.


## **Useful DataFrame Functions**

| Function   | Purpose      |
| ---------- | ------------ |
| head()     | first rows   |
| tail()     | last rows    |
| shape      | dimensions   |
| columns    | column names |
| dtypes     | data types   |
| info()     | dataset info |
| describe() | statistics   |


## **Data Filtering, Sorting, and Handling Missing Values**

**What is Data Filtering?**

**Filtering** means: **selecting only specific rows** based on **conditions**.

**Example**:

- students with marks > 80
- students from CSE department
- students who passed

**Load Dataset**

In [ ]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students_dataset.csv')

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
2         103  Charlie  22.0        CSE    NaN        88.0      Delhi    Yes
3         104    David   NaN         ME   78.0         NaN    Chennai     No
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


**Filtering Rows**

In [ ]:
# Example 1 — Marks Greater Than 80
filtered_df = df[df['Marks'] > 80]

print(filtered_df)

   Student_ID   Name   Age Department  Marks  Attendance       City Passed
0         101  Alice  20.0        CSE   85.0        90.0      Delhi    Yes
1         102    Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
4         105   Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106  Frank  23.0        CSE   88.0        80.0      Delhi    Yes
7         108  Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
9         110   Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


In [ ]:
# Example 2 — Department = CSE
cse_students = df[df['Department'] == 'CSE']

print(cse_students)

   Student_ID     Name   Age Department  Marks  Attendance   City Passed
0         101    Alice  20.0        CSE   85.0        90.0  Delhi    Yes
2         103  Charlie  22.0        CSE    NaN        88.0  Delhi    Yes
5         106    Frank  23.0        CSE   88.0        80.0  Delhi    Yes
8         109      Ian   NaN        CSE   73.0        75.0    NaN     No


## **Multiple Conditions**

**AND Condition**

In [ ]:
result = df[
    (df['Marks'] > 80) &
    (df['Department'] == 'CSE')
]

print(result)

   Student_ID   Name   Age Department  Marks  Attendance   City Passed
0         101  Alice  20.0        CSE   85.0        90.0  Delhi    Yes
5         106  Frank  23.0        CSE   88.0        80.0  Delhi    Yes


**OR Condition**

In [ ]:
result = df[
    (df['Department'] == 'CSE') |
    (df['Department'] == 'ECE')
]

print(result)

   Student_ID     Name   Age Department  Marks  Attendance    City Passed
0         101    Alice  20.0        CSE   85.0        90.0   Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0  Mumbai    Yes
2         103  Charlie  22.0        CSE    NaN        88.0   Delhi    Yes
4         105     Emma  20.0        ECE   92.0        95.0    Pune    Yes
5         106    Frank  23.0        CSE   88.0        80.0   Delhi    Yes
7         108    Helen  22.0        ECE   95.0        92.0  Mumbai    Yes
8         109      Ian   NaN        CSE   73.0        75.0     NaN     No


**Filtering Passed Students**

In [ ]:
passed_students = df[df['Passed'] == 'Yes']

print(passed_students)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
2         103  Charlie  22.0        CSE    NaN        88.0      Delhi    Yes
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


**Selecting Specific Columns After Filtering**

In [ ]:
result = df[df['Marks'] > 80][['Name','Marks']]

print(result)

    Name  Marks
0  Alice   85.0
1    Bob   90.0
4   Emma   92.0
5  Frank   88.0
7  Helen   95.0
9   Jack   81.0


## **Sorting Data**

Sorting arranges data: ascending, descending

**Default**: ascending order

In [ ]:
# Sort by Marks
# Default: ascending order
sorted_df = df.sort_values('Marks')

print(sorted_df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
3         104    David   NaN         ME   78.0         NaN    Chennai     No
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
2         103  Charlie  22.0        CSE    NaN        88.0      Delhi    Yes


In [ ]:
# Descending Order
sorted_df = df.sort_values(
    'Marks',
    ascending=False
)

print(sorted_df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes
3         104    David   NaN         ME   78.0         NaN    Chennai     No
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
2         103  Charlie  22.0        CSE    NaN        88.0      Delhi    Yes


In [ ]:
# Sort by Multiple Columns
sorted_df = df.sort_values(
    ['Department','Marks']
)

print(sorted_df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
2         103  Charlie  22.0        CSE    NaN        88.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
3         104    David   NaN         ME   78.0         NaN    Chennai     No
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


In [ ]:
# Top Students
top_students = df.sort_values(
    'Marks',
    ascending=False
).head(3)

print(top_students)

   Student_ID   Name   Age Department  Marks  Attendance    City Passed
7         108  Helen  22.0        ECE   95.0        92.0  Mumbai    Yes
4         105   Emma  20.0        ECE   92.0        95.0    Pune    Yes
1         102    Bob  21.0        ECE   90.0        85.0  Mumbai    Yes


**Missing Values**

Missing values are: empty cells, **NaN = Not a Number** values

Example:

**Why Missing Values Occur?**

- incomplete data collection
- human error
- corrupted files
- unavailable information

In [ ]:
# Detect Missing Values
print(df.isnull())    # Output: True → missing, False → available

   Student_ID   Name    Age  Department  Marks  Attendance   City  Passed
0       False  False  False       False  False       False  False   False
1       False  False  False       False  False       False  False   False
2       False  False  False       False   True       False  False   False
3       False  False   True       False  False        True  False   False
4       False  False  False       False  False       False  False   False
5       False  False  False       False  False       False  False   False
6       False  False  False       False  False       False  False   False
7       False  False  False       False  False       False  False   False
8       False  False   True       False  False       False   True   False
9       False  False  False        True  False       False  False   False


In [ ]:
# Count Missing Values
print(df.isnull().sum())

Student_ID    0
Name          0
Age           2
Department    1
Marks         1
Attendance    1
City          1
Passed        0
dtype: int64


In [ ]:
# Total Missing Values
print(df.isnull().sum().sum())

6


In [ ]:
# Remove Missing Values
clean_df = df.dropna()

print(clean_df)

   Student_ID   Name   Age Department  Marks  Attendance       City Passed
0         101  Alice  20.0        CSE   85.0        90.0      Delhi    Yes
1         102    Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
4         105   Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106  Frank  23.0        CSE   88.0        80.0      Delhi    Yes
6         107  Grace  21.0         ME   67.0        70.0  Bangalore     No
7         108  Helen  22.0        ECE   95.0        92.0     Mumbai    Yes


In [ ]:
# Fill Missing Values
df['Marks'] = df['Marks'].fillna(0)

In [ ]:
# Fill with Mean
mean_marks = df['Marks'].mean()

df['Marks'] = df['Marks'].fillna(mean_marks)

print(df)

   Student_ID     Name   Age Department  Marks  Attendance       City Passed
0         101    Alice  20.0        CSE   85.0        90.0      Delhi    Yes
1         102      Bob  21.0        ECE   90.0        85.0     Mumbai    Yes
2         103  Charlie  22.0        CSE    0.0        88.0      Delhi    Yes
3         104    David   NaN         ME   78.0         NaN    Chennai     No
4         105     Emma  20.0        ECE   92.0        95.0       Pune    Yes
5         106    Frank  23.0        CSE   88.0        80.0      Delhi    Yes
6         107    Grace  21.0         ME   67.0        70.0  Bangalore     No
7         108    Helen  22.0        ECE   95.0        92.0     Mumbai    Yes
8         109      Ian   NaN        CSE   73.0        75.0        NaN     No
9         110     Jack  24.0        NaN   81.0        89.0  Hyderabad    Yes


In [ ]:
# Fill Missing Age with Average
mean_age = df['Age'].mean()

df['Age'] = df['Age'].fillna(mean_age)

In [ ]:
# Fill Missing Department
df['Department'] = df['Department'].fillna('Unknown')

In [ ]:
# Forward Fill - Copies previous value.
df.fillna(method='ffill')

/tmp/ipykernel_7583/1016775439.py:2: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill')


,Student_ID,Name,Age,Department,Marks,Attendance,City,Passed
0,101,Alice,20.000,CSE,85.0,90.0,Delhi,Yes
1,102,Bob,21.000,ECE,90.0,85.0,Mumbai,Yes
2,103,Charlie,22.000,CSE,0.0,88.0,Delhi,Yes
3,104,David,21.625,ME,78.0,88.0,Chennai,No
4,105,Emma,20.000,ECE,92.0,95.0,Pune,Yes
5,106,Frank,23.000,CSE,88.0,80.0,Delhi,Yes
6,107,Grace,21.000,ME,67.0,70.0,Bangalore,No
7,108,Helen,22.000,ECE,95.0,92.0,Mumbai,Yes
8,109,Ian,21.625,CSE,73.0,75.0,Mumbai,No
9,110,Jack,24.000,Unknown,81.0,89.0,Hyderabad,Yes


In [ ]:
# Checking After Cleaning
print(df.isnull().sum())   # All missing values should become:0

Student_ID    0
Name          0
Age           0
Department    0
Marks         0
Attendance    1
City          1
Passed        0
dtype: int64


# Data Cleaning and Missing Value Treatment Pipeline,

- **Data Cleaning** = Detect Missing Values → Handle Missing Values → Remove Duplicates → Fix Data Types → Get Clean Data

## **1. What is Data Cleaning?**
- Data cleaning means fixing problems in data before analysis.

**Common problems:**

- Missing values
- Duplicate rows
- Wrong data types
- Spelling mistakes
- Outliers

Example

| Name   | Age |
| ------ | --- |
| Ritesh | 25  |
| Aman   | NaN |
| Priya  | 30  |
Here Age is missing for Aman.

## **2. What are Missing Values?**

Missing values are data points that are not available.

In [ ]:
import pandas as pd
import numpy as np

df = pd.DataFrame({
    "Name":["A","B","C"],
    "Age":[20,np.nan,25]
})

print(df)

  Name   Age
0    A  20.0
1    B   NaN
2    C  25.0


NaN means missing value.

## **3. Finding Missing Values**

In [ ]:
df.isnull()

,Name,Age
0,False,False
1,False,True
2,False,False


## **4. Count missing values**

In [ ]:
df.isnull().sum()

,0
Name,0
Age,1


## **4. Remove Missing Values**

Remove Rows

In [ ]:
df.dropna()

,Name,Age
0,A,20.0
2,C,25.0


In [ ]:
# Remove Columns
df.dropna(axis=1)   # axis = 1 means column, # Removes columns containing missing values.


,Name
0,A
1,B
2,C


## **5. Fill Missing Values**

**Instead** of **deleting data**, **fill** it.

- Fill with 0

In [ ]:
# fill with zero
df.fillna(0)

,Name,Age
0,A,20.0
1,B,0.0
2,C,25.0


- Fill with Mean

In [ ]:
df["Age"] = df["Age"].fillna(df["Age"].mean())

## **6. Fill with Median**

In [ ]:
df["Age"] = df["Age"].fillna(df["Age"].median())

Useful when data contains outliers.

## **7. Fill with Mode**

For categorical columns:

In [ ]:
df["Age"] = df["Age"].fillna(df["Age"].mode()[0])

## **8. Remove Duplicate Rows**

In [ ]:
df.drop_duplicates()

,Name,Age
0,A,20.0
1,B,22.5
2,C,25.0


## **9. Check Data Types**

In [ ]:
df.dtypes

,0
Name,object
Age,float64


## **10. Change Data Type**

In [ ]:
df["Age"] = df["Age"].astype(int)

## Complete Missing Value Treatment Pipeline

In [ ]:
import pandas as pd
import numpy as np

df = pd.DataFrame({
    "Name":["A","B","C","D"],
    "Age":[20,np.nan,25,np.nan]
})

print("Before Cleaning")
print(df)

df["Age"] = df["Age"].fillna(df["Age"].mean())

print("\nAfter Cleaning")
print(df)

Before Cleaning
  Name   Age
0    A  20.0
1    B   NaN
2    C  25.0
3    D   NaN

After Cleaning
  Name   Age
0    A  20.0
1    B  22.5
2    C  25.0
3    D  22.5


# **Section 2**

5 objective questions

https://www.menti.com/algxpu2x91fj


**Q1 A data engineer runs df['dept'].astype(str).str.strip().str.title() . What is the result for the entry**
' IT ' ?
- A) 'IT'
- B) 'It'
- C) ' IT '
- D) 'it'

In [ ]:
val = pd.Series([' IT '])
print(repr(val.astype(str).str.strip().str.title().iloc[0]))

'It'


**Q2. In a 500-row dataset, the age column has 80 missing values. Which approach is most appropriate to fill them?**
- A) Drop rows using df.dropna()
- B) Fill with mean age using df['age'].mean()
- C) Fill with median age using np.median(df['age'].dropna())
- D) Fill with string 'unknown' using df['age'].fillna('unknown')

Correct: C

    A Wrong — dropping 80/500 rows is 16% data loss, harmful for a small dataset.
    B Wrong — median is more robust to outliers than mean.
    C Correct — median fill prevents row loss and ignores NaN cleanly.
    D Wrong — a string in a numeric column corrupts the dtype and blocks ML processing.

**Q3. What does pd.to_numeric(salary, errors='coerce') do when it encounters the string "None"?**
- A) Raises a ValueError immediately and stops execution entirely
- B) Converts "None" to NaN instead of raising an error
- C) Leaves the string "None" unchanged inside the numeric Series
- D) Converts "None" to the integer 0 as a built-in default sentinel

In [ ]:
print(pd.to_numeric(pd.Series(['100', 'None', '250']), errors='coerce').tolist())

[100.0, nan, 250.0]


**Q4. A city column has 120 missing entries as None. Which pandas call correctly fills them with 'unknown'?**
- A) df['city'] = df['city'].dropna()
- B) df['city'] = df['city'].astype('unknown')
- C) df['city'] = df['city'].replace('unknown')
- D) df['city'] = df['city'].fillna('unknown')

Correct: D

    A Wrong — dropna() discards null rows, doesn't fill them.
    B Wrong — 'unknown' is not a valid dtype; this errors.
    C Wrong — .replace('unknown') looks for the word 'unknown' — the reverse of what we want.
    D Correct — .fillna('unknown') targets all nulls and assigns the placeholder.



**Q5. Which pandas call should be used to count the exact number of missing entries in every column at once?**
- A) df.head()
- B) df.shape
- C) df.isnull().sum()
- D) df.describe()

Correct: C

    A Wrong — shows the top 5 rows, no aggregation.
    B Wrong — outputs grid dimensions, ignores content.
    C Correct — isnull() flags blanks, .sum() totals them per column.
    D Wrong — describe() gives numeric stats and ignores missing text.



# **Section 3**

Scenario-based practice questions


**Q. Problem Statement:**

A data engineering team at an e-commerce company receives a monthly sales report as a flat list of 12 integers representing unit sales across 4 product categories (rows) and 3 months (columns). A data engineer must write a

**Python function using NumPy that:**
1. Reshapes the flat 12-element array into a (4, 3) matrix.
2. Applies a 10% price-based revenue boost: multiplies each element by the corresponding unit price (provided as
a 1D array of 4 prices, one per product category — broadcast across columns).
3. Computes the per-category total revenue (sum across the 3 months, i.e., axis=1 ).
4. Applies a cap: any category whose total revenue exceeds 5000 is capped at 5000 using np.where() .
5. Returns the final capped per-category revenue array.
Constraints & Requirements:
Use only NumPy ( import numpy as np ); no Pandas or other libraries.
The function signature must be def compute_revenue(sales_flat, unit_prices) .
sales_flat is a 1D NumPy array of exactly 12 non-negative integers.
unit_prices is a 1D NumPy array of exactly 4 positive floats representing price per unit for each of the 4
product categories.
Broadcasting must be used to multiply each row of the reshaped matrix by its corresponding unit price.
np.where() must be used to cap totals at 5000.
Handle the edge case: if sales_flat does not have exactly 12 elements, raise a ValueError with a
descriptive message.

Q2. A **data engineer** at a **fintech startup** has received a raw employee payroll DataFrame containing several data quality issues. The task is to **write a complete pandas cleaning pipeline** that **handles missing values**, **junk characters in the salary column**, **inconsistent capitalization** in the department column, and date parsing — following the strategies for each column type. The **cleaned DataFrame** must have **no remaining null values**, **correct dtypes**, and **pass a validation check**.

**Constraints & Requirements:**
- Use only **pandas** and **NumPy** (import as pd and np )
- For categorical string columns with missing values, **fill with 'unknown'**
- For numeric columns with missing values, **fill with the column's median** computed using **np.median()** on the **non-null values**.
- For the salary column: remove the ₹ symbol and commas using str.replace(regex=False) , then convert to numeric using pd.to_numeric(errors='coerce') , then fill remaining NaN with the median
- For the date column: use pd.to_datetime(errors='coerce', dayfirst=True) and fill NaT with pd.to_datetime('1900-01-01')
- For **text columns:** apply astype(str).str.strip().str.title() to standardize capitalization
- After filling the age column's NaN, cast it to int using astype(int)
- Assemble a clean_df from the individually cleaned columns
- Validate by printing clean_df.isnull().sum() and calling clean_df.info()

In [ ]:
# Use the following 8-row employee dataset. Define it exactly as shown here

import pandas as pd
import numpy as np
data = {
'employee_id': [1, 2, 3, 4, 5, 6, 7, 8],
'name': ['Alice', 'Bob', 'Carol', 'Dave', 'Eve', 'Frank', 'Grace', 'Hank'],'department': ['HR', 'it', None, ' Finance ', 'IT', None, 'hr', 'Finance'],
'age': [25.0, 32.0, None, 28.0, 30.0, None, 27.0, 35.0],
'salary': ['₹45,000', '₹60,000', None, '₹1,20,000', '₹50,000', 'Not Available', '₹55,000', '₹70,000'],
'joining_date':['2022-01-15', '15/03/2021', 'invalid date', '2020-07-01', '10/08/2019', None, '2023-05-20', '2021-11-30'],
'city': ['Mumbai', None, 'Delhi', 'Bangalore', None, 'Chennai', 'Pune', 'Hyderabad']
}
df = pd.DataFrame(data)


In [ ]:
import pandas as pd
import numpy as np

# Sample raw employee payroll DataFrame
df = pd.DataFrame({
    'employee_id': [101, 102, 103, 104, 105],
    'name': ['  john', 'MARY ', None, 'alice', ' Bob '],
    'department': ['finance', 'HR', None, 'it', 'MARKETING'],
    'age': [25, np.nan, 35, np.nan, 40],
    'salary': ['₹50,000', '₹60,000', None, '₹70,500', '₹80,000'],
    'joining_date': ['12/01/2023', '15/02/2022', None,
                     '31/12/2021', 'invalid_date']
})

print("Original DataFrame:")
print(df)

# ----------------------------------------------------
# Step 1: Clean text columns
# ----------------------------------------------------

name_clean = (
    df['name']
    .fillna('unknown')
    .astype(str)
    .str.strip()
    .str.title()
)

department_clean = (
    df['department']
    .fillna('unknown')
    .astype(str)
    .str.strip()
    .str.title()
)

# ----------------------------------------------------
# Step 2: Clean age column
# ----------------------------------------------------

age_median = np.median(df['age'].dropna())

age_clean = (
    df['age']
    .fillna(age_median)
    .astype(int)
)

# ----------------------------------------------------
# Step 3: Clean salary column
# ----------------------------------------------------

salary_clean = (
    df['salary']
    .astype(str)
    .str.replace('₹', '', regex=False)
    .str.replace(',', '', regex=False)
)

salary_clean = pd.to_numeric(
    salary_clean,
    errors='coerce'
)

salary_median = np.median(
    salary_clean.dropna()
)

salary_clean = salary_clean.fillna(
    salary_median
)

# ----------------------------------------------------
# Step 4: Clean joining date column
# ----------------------------------------------------

date_clean = pd.to_datetime(
    df['joining_date'],
    errors='coerce',
    dayfirst=True
)

date_clean = date_clean.fillna(
    pd.to_datetime('1900-01-01')
)

# ----------------------------------------------------
# Step 5: Assemble cleaned DataFrame
# ----------------------------------------------------

clean_df = pd.DataFrame({
    'employee_id': df['employee_id'],
    'name': name_clean,
    'department': department_clean,
    'age': age_clean,
    'salary': salary_clean,
    'joining_date': date_clean
})

# ----------------------------------------------------
# Step 6: Validation
# ----------------------------------------------------

print("\nCleaned DataFrame:")
print(clean_df)

print("\nNull Values Check:")
print(clean_df.isnull().sum())

print("\nDataFrame Information:")
clean_df.info()

Original DataFrame:
   employee_id    name department   age   salary  joining_date
0          101    john    finance  25.0  ₹50,000    12/01/2023
1          102   MARY          HR   NaN  ₹60,000    15/02/2022
2          103    None       None  35.0     None          None
3          104   alice         it   NaN  ₹70,500    31/12/2021
4          105    Bob   MARKETING  40.0  ₹80,000  invalid_date

Cleaned DataFrame:
   employee_id     name department  age   salary joining_date
0          101     John    Finance   25  50000.0   2023-01-12
1          102     Mary         Hr   35  60000.0   2022-02-15
2          103  Unknown    Unknown   35  65250.0   1900-01-01
3          104    Alice         It   35  70500.0   2021-12-31
4          105      Bob  Marketing   40  80000.0   1900-01-01

Null Values Check:
employee_id     0
name            0
department      0
age             0
salary          0
joining_date    0
dtype: int64

DataFrame Information:
<class 'pandas.core.frame.DataFrame'>
RangeIn